### PR TITLE
revise manpage sections

### PR DESCRIPTION
**Pull request by hannesm**
as proposed in #1530

### CODE REVIEW

File: lib_runtime/mirage_runtime.ml [hunk 1]
```diff
@@ -16,8 +16,11 @@
 
 open Cmdliner
 
-let ocaml_section = "OCAML RUNTIME PARAMETERS"
-let unikernel_section = "UNIKERNEL PARAMETERS"
+let s_ocaml = "COMMON OCAML RUNTIME OPTIONS"
+let s_log = "LOG AND MONITORING OPTIONS"
+let s_net = "NETWORK OPTIONS"
+let s_disk = "DISK OPTIONS"
+let s_arg = "UNIKERNEL ARGUMENTS"
```
Comment (reynir) [inline lib_runtime/mirage_runtime.ml:23]:
```suggestion
(* Argument sections are sorted in reverse alphabetic order *)
let s_arg = "UNIKERNEL ARGUMENTS"
let s_net = "NETWORK OPTIONS"
let s_log = "LOG AND MONITORING OPTIONS"
let s_disk = "DISK OPTIONS"
let s_ocaml = "COMMON OCAML RUNTIME OPTIONS"
```

**Review by reynir — APPROVED** [review]

File: lib_runtime/mirage_runtime.ml [hunk 2]
```diff
@@ -16,8 +16,12 @@
 
 open Cmdliner
 
-let ocaml_section = "OCAML RUNTIME PARAMETERS"
-let unikernel_section = "UNIKERNEL PARAMETERS"
+(* Argument sections are sorted in reverse alphabetic order *)
```
Comment (hannesm) [inline lib_runtime/mirage_runtime.ml:19]:
```suggestion
(* These argument sections appear in the manpage in reverse alphabetic order -- as ordered here. *)
```

**Comment by reynir** [comment]
I tested this with unipi (see https://github.com/robur-coop/unipi/pull/29) and I had to pass `~docs:Mirage_runtime.s_args` to `Cmdliner.Arg.info` (which can be done once by defining a local function `let info ?(docs = Mirage_runtime.s_args) = Cmdliner.Arg.info`). Otherwise by default arguments end up in the OPTIONS section near the bottom right before COMMON OPTIONS. The sections are not completely sorted in reverse alphabetic order somehow.

**Comment by hannesm** [comment]
The reverse lexical ordering is for unknown sections -- where name, synopsis, description, commands, arguments, options, common options, exit codes, bugs, .. are known (and have a hardcoded ordering in cmdliner) -- and here the unknown ones are interjected in reverse lexical order after description.

**Comment by reynir** [comment]
I discovered we can control the order of the sections by specifying them in the desired order in the man page:
```diff
diff --git a/lib_runtime/functoria/functoria_runtime.ml b/lib_runtime/functoria/functoria_runtime.ml
index 1a162759..5b49f934 100644
--- a/lib_runtime/functoria/functoria_runtime.ml
+++ b/lib_runtime/functoria/functoria_runtime.ml
@@ -48,7 +48,7 @@ let initialized = ref false
 let help_version = 63
 let argument_error = 64
 
-let with_argv keys s argv =
+let with_argv sections keys s argv =
   let open Cmdliner in
   if !initialized then ()
   else
@@ -68,7 +68,8 @@ let with_argv keys s argv =
         Cmd.Exit.info ~doc:"on OCaml uncaught exception." 255;
       ]
     in
-    match Cmd.(eval_value ~argv (Cmd.v (info ~exits s) t)) with
+    let man = List.map (fun s -> `S s) sections in
+    match Cmd.(eval_value ~argv (Cmd.v (info ~man ~exits s) t)) with
     | Ok (`Ok _) ->
         initialized := true;
         ()
diff --git a/lib_runtime/functoria/functoria_runtime.mli b/lib_runtime/functoria/functoria_runtime.mli
index bbee324f..6ae7611a 100644
--- a/lib_runtime/functoria/functoria_runtime.mli
+++ b/lib_runtime/functoria/functoria_runtime.mli
@@ -41,7 +41,7 @@ val register : 'a Cmdliner.Term.t -> unit -> 'a
 
     [f] will raise [Invalid_argument] if called before cmdliner's evaluation. *)
 
-val with_argv : unit Cmdliner.Term.t list -> string -> string array -> unit
+val with_argv : string list -> unit Cmdliner.Term.t list -> string -> string array -> unit
 (** [with_argv keys name argv] evaluates the [keys] {{!Key.term} terms} on the
     command-line [argv]. [name] is the executable name. On evaluation error the
     application calls [exit(3)] with status [64]. If [`Help] or [`Version] were
diff --git a/lib_runtime/mirage_runtime.ml b/lib_runtime/mirage_runtime.ml
index a7f316ff..07736166 100644
--- a/lib_runtime/mirage_runtime.ml
+++ b/lib_runtime/mirage_runtime.ml
@@ -16,11 +16,11 @@
 
 open Cmdliner
 
-let s_ocaml = "COMMON OCAML RUNTIME OPTIONS"
+let s_ocaml = "OCAML RUNTIME OPTIONS"
 let s_log = "LOG AND MONITORING OPTIONS"
 let s_net = "NETWORK OPTIONS"
 let s_disk = "DISK OPTIONS"
-let s_arg = "UNIKERNEL ARGUMENTS"
+let s_arg = "ARGUMENTS"
 
 type log_threshold = [ `All | `Src of string ] * Logs.level option
 
@@ -249,7 +249,7 @@ let run_leave_iter_hooks () = run leave_iter_hooks
 let at_exit f = add f exit_hooks
 let at_leave_iter f = add f leave_iter_hooks
 let at_enter_iter f = add f enter_iter_hooks
-let with_argv = Functoria_runtime.with_argv
+let with_argv = Functoria_runtime.with_argv [ s_arg; s_net; s_log; s_disk; s_ocaml ]
 let runtime_args = Functoria_runtime.runtime_args
 let register = Functoria_runtime.register
 let argument_error = Functoria_runtime.argument_error
```

**Comment by hannesm** [comment]
oh nice, @reynir -- would you mind to push that change to this PR?

**Comment by hannesm** [comment]
```
-val with_argv : unit Cmdliner.Term.t list -> string -> string array -> unit
+val with_argv : string list -> unit Cmdliner.Term.t list -> string -> string array -> unit
 (** [with_argv keys name argv] evaluates the [keys] {{!Key.term} terms} on the
```

We should describe the new input (and make it an optional argument)!?

**Comment by reynir** [comment]
Please see latest commit. I didn't revert the change of section names, but maybe we should? I don't have any opinions on that.

**Comment by hannesm** [comment]
So, people will forget to use s_arg (which actually we don't need to redefine, we can just use Cmdliner.Manpage.s_arguments).

I propose to make the sections ARGUMENTS OPTTIONS and then our mirage-custom ones. The reasoning is that people will forget, and should nevertheless see their stuff on the top. WDYT?

File: lib_runtime/functoria/functoria_runtime.ml [hunk 3]
```diff
@@ -48,7 +48,7 @@ let initialized = ref false
 let help_version = 63
 let argument_error = 64
 
-let with_argv keys s argv =
+let with_argv sections keys s argv =
```
Comment (hannesm) [inline lib_runtime/functoria/functoria_runtime.ml:51]:
```suggestion
let with_argv ?sections keys s argv =
```

File: lib_runtime/functoria/functoria_runtime.ml [hunk 4]
```diff
@@ -68,7 +68,8 @@ let with_argv keys s argv =
         Cmd.Exit.info ~doc:"on OCaml uncaught exception." 255;
       ]
     in
-    match Cmd.(eval_value ~argv (Cmd.v (info ~exits s) t)) with
+    let man = List.map (fun s -> `S s) sections in
+    match Cmd.(eval_value ~argv (Cmd.v (info ~man ~exits s) t)) with
```
Comment (hannesm) [inline lib_runtime/functoria/functoria_runtime.ml:72]:
```suggestion
    let man = Option.map (fun s -> `S s) sections in
    match Cmd.(eval_value ~argv (Cmd.v (info ?man ~exits s) t)) with
```

File: lib_runtime/functoria/functoria_runtime.mli [hunk 5]
```diff
@@ -41,11 +41,13 @@ val register : 'a Cmdliner.Term.t -> unit -> 'a
 
     [f] will raise [Invalid_argument] if called before cmdliner's evaluation. *)
 
-val with_argv : unit Cmdliner.Term.t list -> string -> string array -> unit
-(** [with_argv keys name argv] evaluates the [keys] {{!Key.term} terms} on the
-    command-line [argv]. [name] is the executable name. On evaluation error the
-    application calls [exit(3)] with status [64]. If [`Help] or [`Version] were
-    evaluated, [exit(3)] is called with status [63]. *)
+val with_argv : string list -> unit Cmdliner.Term.t list -> string -> string array -> unit
+(** [with_argv sections keys name argv] evaluates the [keys] {{!Key.term} terms} on the
```
Comment (hannesm) [inline lib_runtime/functoria/functoria_runtime.mli:45]:
```suggestion
val with_argv : ?sections:string list -> unit Cmdliner.Term.t list -> string -> string array -> unit
(** [with_argv ~sections keys name argv] evaluates the [keys] {{!Key.term} terms} on the
```

File: lib_runtime/mirage_runtime.ml [hunk 6]
```diff
@@ -16,8 +16,12 @@
 
 open Cmdliner
 
-let ocaml_section = "OCAML RUNTIME PARAMETERS"
-let unikernel_section = "UNIKERNEL PARAMETERS"
+(* These argument sections appear in the manpage in reverse alphabetic order -- as ordered here. *)
+let s_arg = "UNIKERNEL ARGUMENTS"
```
Comment (hannesm) [inline lib_runtime/mirage_runtime.ml:20]:
```suggestion
```

File: lib_runtime/mirage_runtime.ml [hunk 7]
```diff
@@ -16,8 +16,12 @@
 
 open Cmdliner
 
-let ocaml_section = "OCAML RUNTIME PARAMETERS"
-let unikernel_section = "UNIKERNEL PARAMETERS"
+(* These argument sections appear in the manpage in reverse alphabetic order -- as ordered here. *)
```
Comment (hannesm) [inline lib_runtime/mirage_runtime.ml:19]:
```suggestion
```

File: lib_runtime/mirage_runtime.mli [hunk 8]
```diff
@@ -42,6 +42,22 @@ module Conv : sig
   (** [allocation_policy] converts allocation policy. *)
 end
 
+(** {2 Manpage sections} *)
+val s_arg : string
+(** [s_arg] is suggested to be used for your custom unikernel arguments. *)
+
```
Comment (hannesm) [inline lib_runtime/mirage_runtime.mli:48]:
```suggestion
```

File: lib_runtime/mirage_runtime.ml [hunk 9]
```diff
@@ -249,7 +250,7 @@ let run_leave_iter_hooks () = run leave_iter_hooks
 let at_exit f = add f exit_hooks
 let at_leave_iter f = add f leave_iter_hooks
 let at_enter_iter f = add f enter_iter_hooks
-let with_argv = Functoria_runtime.with_argv
+let with_argv = Functoria_runtime.with_argv [ s_arg; s_net; s_log; s_disk; s_ocaml ]
```
Comment (hannesm) [inline lib_runtime/mirage_runtime.ml:253]:
```suggestion
let with_argv = Functoria_runtime.with_argv ~sections:[ Cmdliner.Manpage.s_arguments ; Cmdliner.Manpage.s_options; s_net; s_log; s_disk; s_ocaml ]
```

File: lib_runtime/mirage_runtime.ml (hunk 7 above)
Comment (reynir) [reply lib_runtime/mirage_runtime.ml:19]: I swear I edited that comment :/ must have lost the commit when rebasing somehow...

Comment (reynir) [reply lib_runtime/mirage_runtime.ml]: Ah it was uncommitted and now part of dc796e36c4b67381525123237762a512555e5616

**Comment by hannesm** [comment]
I tested this with the above linked unipi PR, and to me this is a great improvement.

Especially that the default section for optional arguments and required arguments (if you do not specify `~docs` in your `Cmdliner.Arg.info`) are shown at the top in the manpage -- so users won't need to search for these.

Any other opinions?

**Comment by samoht** [comment]
Looks great, thanks!

**Pull request merged**